mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td

```diff
@@ -1034,6 +1034,37 @@ def LLVM_TBAATagArrayAttr
   let constBuilderCall = ?;
 }
 
+//===----------------------------------------------------------------------===//
+// ConstantRangeAttr
+//===----------------------------------------------------------------------===//
+def LLVM_ConstantRangeAttr : LLVM_Attr<"ConstantRange", "constant_range"> {
+  let parameters = (ins
+    "IntegerAttr":$lower,
+    "IntegerAttr":$upper
+  );
+  let summary = "A range of two integers, corresponding to LLVM's ConstantRange";
+  let description = [{
+    A pair of two integers, mapping to the ConstantRange structure in LLVM IR,
+    which is allowed to wrap or be empty.
+
+    The range represented is [Lower, Upper), and is either signed or unsigned
+    depending on context.
+
+    `lower` and `upper` must have the same width.
+  }];
+
+  let builders = [
+    AttrBuilder<(ins "uint32_t":$bitWidth, "int64_t":$lower, "int64_t":$upper)>
```

Comment:
Can we just store this as `<uint32_t, int64_t, int64_t>` instead of using IntegerAttr?

Reply:
I'd argue no - a range of i128s is permitted even though we currently don't have usecases

Reply:
OK, but then this builder is incorrect. It is also leading to inneficiency by hiding costly operations from the user.
 
(I'm also not sure when a i64 would ever be too small for any range in the compiler?)

Reply:
I mean, we *could* go for `<APInt, APInt>` ... and I'd say the builder isn't incorrect, it's a convenience for the typical case where you'd want `<APInt(width, lower), APINt(width, upper)>`

Reply:
> I mean, we could go for <APInt, APInt> 

That would be a bit less "sparse" than the current representation indeed, and it would also encode the bitWidth once.

> it's a convenience for the typical case where you'd want <APInt(width, lower), APINt(width, upper)>

You're missing the restriction on the bitwidth right now I believe: the builder is taking int64_t.

Reply:
In further notes:

The `VScaleRangeAttr` elsewhere is this file is also defined as `<IntegerAttr, IntegerAttr>` - so using a pair of attributes for a pair of APInt is established practice in this file.

Also, a pair of APInt wouldn't save encoding the bitwidth twice - both APInt values need the width.

And, last I checked, none of the `(uint32_t bitwidth, int64_t value)`-type builders actually check for the width being under 64 bits (partly because `IntegerAttr(context, /*bitwidth=*/128, /*value=*/0)` is legitimate, and that translates to the equivalent `ConstantRanges` builder)

Reply:
>  so using a pair of attributes for a pair of APInt is established practice in this file.

Sure, there are tons of bad practices in the codebase, it's rarely an argument to continue doing so though.

> And, last I checked, none of the (uint32_t bitwidth, int64_t value)-type builders actually check for the width being under 64 bits

I don't understand: there is nothing to check, my comment was that the `int64_t value` can't represent values more than 64 bits...

Reply:
Re not understanding my second note, then I don't understand what you mean by "you're missing" in
> You're missing the restriction on the bitwidth right now I believe: the builder is taking int64_t.